### PR TITLE
Auth- II - Joseph Bradley - Braden Walker

### PR DESCRIPTION
**Pull request by jmbradley**

### CODE REVIEW

**Comment by jmbradley** [comment]
Wrapping client side auth

File: index.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+//       tablename: 'sessions',
+//       sidefieldname: 'sid',
+//       knex: db,
+//       createtable: true,
+//       clearIntervale: 1000 * 60 * 60,
+//     }),
+// };
+
+
+const jwtSecret = 'quikbrownfox';
```
Comment (Bradiowave) [inline index.js]: Put secrets in environment variables so they aren't visible on github. Even though it's not as important on this project, it's a good thing to make a habit of.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+};
+
+server.use(
+  express.json(), 
+  cors(),
+//   session(sessionConfig),
+);
+
+server.post('/api/register', (req, res) => {
+  const credentials = req.body;
```
Comment (Bradiowave) [inline index.js]: Destructure the username, password, and department from the request body for database security purposes.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  const hash = bcrypt.hashSync(credentials.password, 14);
+  credentials.password = hash;
+
+  db('users').insert(credentials).then(ids => {
+    const id = ids[0];
+    // req.session.username = credentials.username;  
+    res.status(201).json({ newUserId: id})
+  })
+   .catch(err => {
+     res.status(500).json(err);
```
Comment (Bradiowave) [inline index.js]: Write an error message explaining why the user couldn't register and how they can solve the issue.

**Comment by Bradiowave** [comment]
## MVP
- Build a User table that includes
    - [x] username
    - [x] password
    - [ ] department (string used to group the users)
- [x] Use JSON Web Tokens to keep users authenticated
- Design and build the following endpoints
    - [x] POST /api/register
        - [x] Hash the password before saving the user
    - [x] POST /api/login
        - [x] Use the credentials sent inside the body to authenticate the user. On successful login, create a new JWT with the user id as the subject and send it back to the client. 
        - [x] If login fails, respond with the correct status code and the message: 'You shall not pass!'
    - [x] GET /api/users
        - [x] If the user is logged in, respond with an array of all the users contained in the database. 
        - [x] If the user is not logged in repond with the correct status code and the message: 'You shall not pass!'.
- Implement a React Client
    - [x] use create-react-app to generate a application to server as the client for the Web API.
    - [x] inside the React application add client-side routes and components for signup, signin and showing the list of users stored in the database.
    - [ ] the /signup route should provide a form to gather username, password and department for the user and make a POST request to the /api/register route on the API. If the user is created successfully, take the returned token, save it to the browser's local storage and redirect the user to the /users route, where they should see the list of users.
    - [x] the /signin route should provide a form to gather username and password for the user and make a POST request to the /api/login route on the API. Upon successful login, persist the returned token to the browser's local storage and redirect the user to the /users route.
    - [x] the /users route should read the token from local storage and make a GET request to the /api/users route on the API attaching the token as the value of the Authorization header.
    - [x] provide a button to sign out that will remove the token from local storage.

## Stretch
- [ ] add the code necessary so that when a client makes a GET request to /api/users the server only returns documents with the same department as the logged in user.
- [ ] add any extra functionality to make the application more user friendly like showing a message and redirecting to /signin if an unauthenticated user tries to access the list of users in the /users route

## What Looks Great
- MVP nearly complete! (only missing the signup component and the department column on each user)
- React app styling looks nice
- Frequent and descriptive commits
- Code is mostly well formatted and easy to read
- Great file structure. React components contained in appropriate directories. Database migrations also contained in a folder.

## What Could Be Improved
- Add a signup component in the React app, and a department column to your database per the README.
- Follow this guide to make your commit messages even better (https://chris.beams.io/posts/git-commit/)
- Some of the comments are distracting and unnecessary to the code. You can keep them in if they help you learn, but know that theses shouldn't usually be kept in.

## Rating (1-3)
**2**

**Pull request closed without merging**